packages/utilities/src/FocusRectsProvider.tsx

```diff
@@ -0,0 +1,65 @@
+import * as React from 'react';
+import { FocusRectsContext } from './useFocusRects';
+
+export type FocusRectsProviderParams = {
+  /**
+   * Ref to the root element that this is providing focus rects for.
+   */
+  providerRef: React.RefObject<HTMLElement>;
+
+  /**
+   * Indicates that this is the root of a layer, and should not inherit the providerRef from a parent context.
+   */
+  layerRoot?: boolean;
+};
+
+export const FocusRectsProvider: React.FC<FocusRectsProviderParams> = params => {
+  const { providerRef, layerRoot } = params;
+  const [registeredProviders] = React.useState<HTMLElement[]>([]);
+  const parentContext = React.useContext(FocusRectsContext);
+
+  // Inherit the parent context if it exists, unless this is a layer root.
+  // This allows the topmost provider element in the DOM tree to handle the focus events.
+  // Since layers are in a separate HTML tree from their parent, they shouldn't use the parent's providerRef.
+  const inheritParentContext = parentContext !== undefined && !layerRoot;
+
+  const context = React.useMemo(
+    () =>
+      inheritParentContext
+        ? undefined
+        : {
+            providerRef,
+            registeredProviders,
+            registerProvider: (e: HTMLElement) => {
+              // Register this child provider with the current context, and any parent contexts
+              registeredProviders.push(e);
```

Comment:
It looks like this is written this way to avoid a re-render of all children of this Context when `registeredProviders` is changed. Seeing as how this value is read in response to a mouse/pointer event I think it should be safe to do this way.

Reply:
Yeah, Sean described it well. 

It's not very react-y, but the whole way that this handles focus rect visibility is not very react-y, so I was trying to preserve the current behavior as much as possible, while adding the new functionality to have it update child providers when the parent is updated.